config/initializers/rack_attack.rb

```diff
@@ -20,25 +20,48 @@ class Rack::Attack
                              '/downloads/')
   end
 
+  # Allow authenticated user
+  Rack::Attack.safelist('authenticated user') do |req|
+    !req.env['warden'].user.blank?
+  end
+
   # Based on occurrences during the week of June 16, 2019, DigitalHub
   # suffered a DDoS-like event. Inspection of logs during the event showed
   # +10k requests per hour to digitalhub, from thousands of different ips.
   # Throttling by ip would prove ineffective in this case. Though in the
   # hundreds, the ips showed a subnet-like pattern. The following throttle
   # rules, tracking by subnet, are based off this event.
-  catalog_limit_proc = proc { |req| req.env['warden'].user.blank? ? 100 : 5000 }
+
   # Throttle /catalog by subnet
-  # 5000/hr for authenticated, 100/hr for unauthenticated
-  throttle('catalog/subnet', limit: catalog_limit_proc, period: 1.hour) do |req|
-    if req.fullpath.start_with?('/catalog?', '/catalog/') && !req.fullpath.start_with?('/catalog?page')
-      req.ip.slice(0..req.ip.rindex("."))
+  # Exponential Backoff for unauthenticated user
+  #   Level 6: 96 requests per 64 minutes
+  #   Level 7: 112 requests per 128 minutes
+  #   Level 8: 128 requests per 256 minutes
+  #   Level 9: 144 requests per 512 minutes
+  #   Level 10: 160 requests per 1024 minutes
+  (6..10).each do |level|
+    # We set a strict limit for the catalog path because facets/filters consume
+    # a lot of resources, hence slowing down digitalhub. Navigating around the
+    # '/catalog?page' path is fine.
+    throttle("catalog/subnet/#{level}", :limit => (16 * level), :period => (2 ** level).minutes) do |req|
```

Comment:
What is the meaning of `level` here?

The exponential part is done by us. Does `throttle` do the backoff part?

Reply:
From levels 6-10, we set different throttle rules for each. For each level, we decrease the requests rate limit, and increase the backoff period. Does that help? I can explain in person if easier?